fmn/formatters.py

```diff
@@ -548,43 +551,50 @@ def email_confirmation(confirmation):
         support_email=config.app_conf['fmn.support_email'],
         username=confirmation.openid,
     ).strip()
-    email_message.set_payload(content)
+    email_message.set_payload(content.encode('utf-8'), 'utf-8')
     return email_message.as_string()
 
 
-def _base_email(recipient=None, messages=None):
+def _add_fmn_headers(email_message, recipient=None, messages=None):
     """
-    Create an email Message with some basic headers to mark the email as auto-generated.
+    Add basic headers to mark the email as auto-generated.
+    If a recipient is provided, add it to 'To' header.
+    If a list of messages is provided, add related custom FMN headers.
+
+    Args:
+        email_message (email.message.Message): The email message to customize
+        recipient (dict): A dictionary containing (at a minimum) the
+            `email_address` (str) and `triggered_by_links` (bool) keys.
+        messages (list): A list of dictionaries, where each dictionary is a fedmsg.
 
     Returns:
-        email.Message.Message: The email message object with the 'Precedence' and 'Auto-Submitted'
-            headers set.
+        email.message.Message: The email message object with basic and custom headers set.
     """
-    email_message = email_module.Message.Message()
+    email_message['From'] = config.app_conf['fmn.email.from_address']
+    if recipient and 'email address' in recipient:
+        email_message['To'] = recipient['email address']
+
     # Although this is a non-standard header and RFC 2076 discourages it, some
     # old clients don't honour RFC 3834 and will auto-respond unless this is set.
     email_message.add_header('Precedence', 'Bulk')
     # Mark this mail as auto-generated so auto-responders don't respond; see RFC 3834
     email_message.add_header('Auto-Submitted', 'auto-generated')
-    email_message.add_header('From', config.app_conf['fmn.email.from_address'].encode('utf-8'))
-
-    if recipient and 'email address' in recipient:
-        email_message.add_header('To', recipient['email address'].encode('utf-8'))
 
     if messages:
         try:
             topics = set([message['topic'] for message in messages])
             for topic in topics:
-                email_message.add_header('X-Fedmsg-Topic', topic.encode('utf-8'))
+                email_message.add_header('X-Fedmsg-Topic', topic)
             categories = set([topic.split('.')[3] for topic in topics])
             for cat in categories:
-                email_message.add_header('X-Fedmsg-Category', cat.encode('utf-8'))
+                email_message.add_header('X-Fedmsg-Category', cat)
         except Exception:
             _log.exception('Unable to parse message topic and category for %r', messages)
 
         usernames = set()
         packages = set()
         for msg in messages:
+            email_message.add_header('X-Fedmsg-Id', msg['msg_id'])
```

Comment:
Sorry, somehow Github stopped sending me emails, so I've just see your review...
I will rebase the commits and give a look to the maximum headers number ASAP.

Thanks

Reply:
About X-Fedmsg-Id header: if the recipient is set to 'verbose', the new digest is made of many email messages, each one with its own header ID.

The only problem is when the recipient is set to 'not verbose', in this case the batch_email is not a real digest, but a self made summary like the old one. So in this case we can have a single email with up to 1000 headers ID.

I can't find any limit to the number of headers an email can carry in the standards. Also I don't think the solution applied to the package header would be appropriate in this case. I'm open to suggestions.

Reply:
That limit is a header size limit, which is configured on pretty much all MTAs.
A concrete example: http://www.postfix.org/postconf.5.html, search for `header_size_limit (default: 102400)`.
Technically we could increase it for our mail server, but that only increases the chance that it gets refused by the recipients mailserver.

Reply:
If so, we have troubles with the other X-Fedmsg headers as well. Each message adds:
- one X-Fedmsg-Topic
- one or more X-Fedmsg-Category
- one X-Fedmsg-Id
- one or more (up to ten) X-Fedmsg-Package
- one X-Fedmsg-Num-Packages
- one or more X-Fedmsg-Username

With an `header_size_limit` of 102400 even limiting the recap to 250 messages (the maximum that can be set on the web UI) can easily overwhelm the headers limit.

This can only impact on the 'recap' email, not on the real digest email, because in the digest the X-Fedmsg headers are in the attached messages.
So I would propose to not add the X-Fedmsg headers in the recap at all... but, since I don't know exactly the real word use case of those headers, I probably be wrong and break some other automated tool. Note that also setting the X-Fedmsg headers inside the attached messages instead of setting them in the top level digest can break those tools.

Reply:
Sorry, I let this slip through the cracks. I *think* I'd be okay without including the X-Fedmsg headers in recaps. They make some amount of sense for filtering individual emails, but probably not with a huge digest.